out_dynatrace: add configuration for Dynatrace based on http plugin (#1461)
fluent/fluent-bit-docs: pipeline/outputs/dynatrace.md

---
description: Send logs to Dynatrace
---

# Dynatrace

> Stream logs to [Dynatrace](https://www.dynatrace.com) by utilizing the **http** plugin to send data to [Dynatrace generic log ingest API](https://docs.dynatrace.com/docs/shortlink/lma-generic-log-ingestion).
> 
> Before starting you need to get a [Dynatrace API](https://docs.dynatrace.com/docs/shortlink/api-authentication) token with the `logs.ingest` (Ingest Logs) scope, and determine your [environment ID](https://docs.dynatrace.com/docs/shortlink/monitoring-environment#environment-id).  Then configure **http** output plugin with the following HTTP parameters.

Reviewer: I implemented line wrapping as part of my suggestion and added a "the" for clarity.

Also, I removed the second line, as it was unneeded. It just restated the process to be used in the getting started section. Stating things more than once in the docs just means we have two sources of information that have to be maintained, in the case things change.

```suggestion
Stream logs to [Dynatrace](https://www.dynatrace.com) by utilizing the **http**
plugin to send data to the
[Dynatrace generic log ingest API](https://docs.dynatrace.com/docs/shortlink/lma-generic-log-ingestion).
```

> ## Configuration Parameters
> 
> | Key                        | Description                                                                                                                                                                                                                         | default                                            |
> | -------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------- |
> | header                     | The specific header for content-type                                                                                                                                                                                                | Content-Type application/json; charset=utf-8       |
> | header                     | The specific header for authorization token, where {your-API-token-here} is the Dynatrace API token with log ingest scope                                                                                                           | Authorization Api-Token {your-API-token-here}      |
> | Allow_Duplicated_Headers   | Specifies duplicated header use                                                                                                                                                                                                     |  false                                             |
> | host                       | Your Dynatrace environment hostname where {your-environment-id} is your environment ID.                                                                                                                                             | {your-environment-id}.live.dynatrace.com           |
> | port                       | TCP port of your Dynatrace host                                                                                                                                                                                                     | 443                                                |
> | uri                        | Specify the HTTP URI for Dynatrace log ingest API                                                                                                                                                                                   | /api/v2/logs/ingest                                |
> | format                     | The data format to be used in the HTTP request body                                                                                                                                                                                 | json                                               |
> | json_date_format           | Date format standard for JSON                                                                                                                                                                                                       | iso8601                                            |
> | json_date_key              | Fieldname specifying message timestamp                                                                                                                                                                                              | timestamp                                          |
> | tls                        | Specify to use TLS                                                                                                                                                                                                                  | on                                                 |
> | tls.verify                 | TLS verification                                                                                                                                                                                                                    | on                                                 |

Reviewer: I reformatted the table to help with content display when editing, and fixed some formatting issues.

Also, did you mean to list the header key twice?

```suggestion
## Configuration Parameters

| Key                        | Description | Default |
| -------------------------- | ----------- | ------- |
| header                     | The specific header for content-type. | `Content-Type application/json; charset=utf-8` |
| header                     | The specific header for authorization token, where {your-API-token-here} is the Dynatrace API token with log ingest scope. | Authorization Api-Token {your-API-token-here} |
| Allow_Duplicated_Headers   | Specifies duplicated header use. | `false` |
| host                       | Your Dynatrace environment hostname where {your-environment-id} is your environment ID. | `{your-environment-id}.live.dynatrace.com` |
| port                       | TCP port of your Dynatrace host. | `443` |
| uri                        | Specify the HTTP URI for Dynatrace log ingest API. | `/api/v2/logs/ingest` |
| format                     | The data format to be used in the HTTP request body. | `json` |
| json_date_format           | Date format standard for JSON. | `iso8601` |
| json_date_key              | Fieldname specifying message timestamp. | `timestamp` |
| tls                        | Specify to use TLS. | `on` |
| tls.verify                 | TLS verification. | `on` |
```

Author: Thanks for the edits!
Yes the header key should be used twice. Matching the guide here: https://docs.dynatrace.com/docs/shortlink/lma-stream-logs-with-fluent-bit#customize-fluent-bit-configuration

> ## Getting Started
> 
> 1. Get a [Dynatrace API](https://docs.dynatrace.com/docs/shortlink/api-authentication) token with the `logs.ingest` (Ingest Logs) scope.
> 2. Determine your Dynatrace [environment ID](https://docs.dynatrace.com/docs/shortlink/monitoring-environment#environment-id).
> 3. In your main Fluent Bit configuration file, append the following _Output_ section:
> 
> ```text
> [OUTPUT]
>     name         http
>     match        *
>     header       Content-Type application/json; charset=utf-8
>     header       Authorization Api-Token {your-API-token-here}
>     allow_duplicated_headers false
>     host         {your-environment-id}.live.dynatrace.com
>     Port         443
>     URI          /api/v2/logs/ingest
>     Format       json
>     json_date_format iso8601
>     json_date_key timestamp
>     tls          On
>     tls.verify   On
> ```
> 
> 4. Once the configuration is applied and new logs are appended, verify their receipt within Dynatrace. This can be done using Dynatrace [Notebooks](https://docs.dynatrace.com/docs/observe-and-explore/dashboards-and-notebooks/notebooks) with a Dynatrace Query Language query, for example:
> 
>     ```
>     fetch logs
>     ```

Reviewer: This set of changes puts in line wraps (to help with formatting the raw content), renumbers the lines (Markdown supports using all 1s, which will autonumber lists), and implements a few more minor changes.

```suggestion
## Getting started

To get started with sending logs to Dynatrace:

1. Get a [Dynatrace API](https://docs.dynatrace.com/docs/shortlink/api-authentication)
   token with the `logs.ingest` (Ingest Logs) scope.
1. Determine your Dynatrace
   [environment ID](https://docs.dynatrace.com/docs/shortlink/monitoring-environment#environment-id).
1. In your main Fluent Bit configuration file, append the following `Output` section:

   ```text
   [OUTPUT]
       name         http
       match        *
       header       Content-Type application/json; charset=utf-8
       header       Authorization Api-Token {your-API-token-here}
       allow_duplicated_headers false
       host         {your-environment-id}.live.dynatrace.com
       Port         443
       URI          /api/v2/logs/ingest
       Format       json
       json_date_format iso8601
       json_date_key timestamp
       tls          On
       tls.verify   On
   ```

1. After the configuration is applied and new logs are appended, verify their receipt
   within Dynatrace. This can be done using Dynatrace
   [Notebooks](https://docs.dynatrace.com/docs/observe-and-explore/dashboards-and-notebooks/notebooks)
   with a Dynatrace Query Language query. For example:

   ```bash
   fetch logs
   ```
```

## References

> * [Dynatrace Fluent Bit documentation](https://docs.dynatrace.com/docs/shortlink/lma-stream-logs-with-fluent-bit)
> * [Fluent Bit integration in Dynatrace Hub](https://www.dynatrace.com/hub/detail/fluent-bit/?filter=log-management-and-analytics) 
> * [Video: Stream a Log File to Dynatrace using Fluent Bit](https://www.youtube.com/watch?v=JJJNxhtJ6R0)
> * [Blog: Easily stream logs from Fluent Bit to Dynatrace](https://www.dynatrace.com/news/blog/easily-stream-logs-with-fluent-bit-to-dynatrace/)

Reviewer: For clarity, we encourage the use of hyphens for bullets instead of asterisks, as it helps avoid Markdown scanning confusion with bold codes.

```suggestion
- [Dynatrace Fluent Bit documentation](https://docs.dynatrace.com/docs/shortlink/lma-stream-logs-with-fluent-bit)
- [Fluent Bit integration in Dynatrace Hub](https://www.dynatrace.com/hub/detail/fluent-bit/?filter=log-management-and-analytics) 
- [Video: Stream a Log File to Dynatrace using Fluent Bit](https://www.youtube.com/watch?v=JJJNxhtJ6R0)
- [Blog: Easily stream logs from Fluent Bit to Dynatrace](https://www.dynatrace.com/news/blog/easily-stream-logs-with-fluent-bit-to-dynatrace/)
```
